zwave_js_server/model/node/__init__.py

```diff
@@ -332,70 +344,75 @@ def firmware_update_progress(self) -> NodeFirmwareUpdateProgress | None:
     @property
     def highest_security_class(self) -> SecurityClass | None:
         """Return highest security class configured on the node."""
-        if (security_class := self.data.get("highestSecurityClass")) is None:
+        if (security_class := self._data.get("highestSecurityClass")) is None:
             return None
         return SecurityClass(security_class)
 
     @property
     def is_controller_node(self) -> bool:
         """Return whether the node is a controller node."""
-        return self.data["isControllerNode"]
+        return self._data["isControllerNode"]
 
     @property
     def keep_awake(self) -> bool:
         """Return whether the node is set to keep awake."""
-        return self.data["keepAwake"]
+        return self._data["keepAwake"]
 
     def update(self, data: NodeDataType) -> None:
         """Update the internal state data."""
-        self.data = data
-        self._device_config = DeviceConfig(self.data.get("deviceConfig", {}))
+        self._data = copy.deepcopy(data)
+        self._device_config = DeviceConfig(self._data.get("deviceConfig", {}))
         self._statistics = NodeStatistics(
-            self.client, self.data.get("statistics", DEFAULT_NODE_STATISTICS)
+            self.client, self._data.get("statistics", DEFAULT_NODE_STATISTICS)
         )
 
-        # Remove stale values
-        value_ids = (_get_value_id_str_from_dict(self, val) for val in data["values"])
-        self.values = {
-            value_id: val
-            for value_id, val in self.values.items()
-            if value_id in value_ids
+        new_values_data = {
+            _get_value_id_str_from_dict(self, val): val
+            for val in self._data.pop("values")
         }
+        stale_value_ids = list(set(self.values) - set(new_values_data))
```

Comment:
We don't need to create lists.
```suggestion
        stale_value_ids = set(self.values) - set(new_values_data)
```